stl/inc/xutility

```diff
@@ -4650,55 +4650,51 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-                                                && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2>>,
-          negation<is_volatile<_Elem1>>, negation<is_volatile<_Elem2>>,
-          // order matters here: being integral is a precondition of _Value_equality_is_bitwise_equality
-          _Value_equality_is_bitwise_equality<_Elem1, _Elem2>>> {
-    // allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2,
+    enable_if_t<sizeof(_Elem1) == sizeof(_Elem2)
+             && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2>
+             && !is_volatile_v<_Elem1> && !is_volatile_v<_Elem2>, equal_to<>>
+    > = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
+// clang-format on
 
+// allow memcmping pointers-to-cv-T with equal_to<>
 template <class _Elem1, class _Elem2>
-struct _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>>
-    : is_same<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>::type {}; // allow memcmping pointers-to-cv-T with equal_to<>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>> =
+    is_same_v<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>;
 
 #ifdef __cpp_lib_byte
+// allow memcmping std::byte
 template <>
-struct _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> : true_type {}; // allow memcmping std::byte
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> = true;
 #endif // __cpp_lib_byte
 
+// treat equal_to with exact T as equal_to<> this is safe because we only activate the optimization for builtin
+// _Elem (and std::byte)
 template <class _Elem>
-struct _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>>
-    : _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>::type {
-    // treat equal_to with exact T as equal_to<> this is safe because we only activate the optimization for builtin
-    // _Elem (and std::byte)
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>> =
+    _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>;
 
-template <class _Iter1, class _Iter2, class _Pr>
-false_type _Equal_memcmp_is_safe(const _Iter1&, const _Iter2&, const _Pr&) {
-    // return equal optimization category for arbitrary iterators
-    return {};
-}
+// return equal optimization category for arbitrary iterators
+template <class _Elem1, class _Elem2, class _Pr>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe = false;
 
-template <class _Obj1, class _Obj2, class _Pr>
-typename _Equal_memcmp_is_safe_helper<remove_const_t<_Obj1>, remove_const_t<_Obj2>, _Pr>::type _Equal_memcmp_is_safe(
-    _Obj1* const&, _Obj2* const&, const _Pr&) { // return equal optimization category for pointers
-    return {};
-}
+// return equal optimization category for pointers
+template <class _Elem1, class _Elem2, class _Pr>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe<_Elem1*, _Elem2*, _Pr> =

```

Comment:
This is introducing a potential usability issue. Previously, `_Equal_memcmp_is_safe(ptr1, ptr2, pred)` ignored top-level `const` on the pointers, due to `_Obj1* const&, _Obj2* const&`. Now, `_Equal_memcmp_is_safe<_Elem1*, _Elem2*, _Pr>` won't activate for `int* const`. This could be formed if an algorithm applies top-level `const` to an unwrapped iterator and then uses `decltype` which will pick up that `const`.

Should we accept top-level `const` here, and add tests to `VSO_0180469_ptr_cat`? Long ago, I would have introduced another layer of metaprogramming to handle this - but it should be sufficient to use your `enable_if_t` trick here, enabling this specialization for `is_pointer_v` (the primary type categories ignore top-level const) and then using `remove_pointer_t` before `remove_const_t` below.

Reply:
I will have to think about this a bit more. I thought that the `remove_const_t` was sufficient but it seems it is not.

In any case I should add some tests.

Reply:
The `remove_const_t` there is ignoring constness on the elements, but not top-level const on the pointers.

Reply:
So I wrote some small tests

```cpp
    // Top level const should work
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1*, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, const Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1*, const Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, const Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1*, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, const Elem2* const, equal_to<>>();
```

However, I did not get it to pass yet. Is there some secret way to ask the compiler which templates are instantiated?

Reply:
The technique I use is getting a template to instantiate some other template that I've declared but not defined. For example:

```cpp
template <typename A, typename B, typename C> struct Meow;

template <typename T, typename U> auto add(T t, U u) -> decltype(t + u) {
    Meow<T, U, decltype(t + u)>::print();
    return t + u;
}

int main() {
    add(static_cast<unsigned char>(1), static_cast<unsigned short>(2));
}
```

MSVC emits:
```
error C2027: use of undefined type 'Meow<T,U,int>'
        with
        [
            T=unsigned char,
            U=unsigned short
        ]
```

Clang emits:
```
error: implicit instantiation of undefined template 'Meow<unsigned char, unsigned short, int>'
```

Reply:
Thanks to your help this is now working